charms/kfp-profile-controller/src/charm.py

```diff
@@ -4,319 +4,153 @@
-                                        # {
-                                        #     "apiVersion": "networking.istio.io/v1alpha3",
-                                        #     "resource": "destinationrules",
-                                        #     "updateStrategy": {"method": "InPlace"},
-                                        # },
-                                        # {
-                                        #     "apiVersion": "security.istio.io/v1beta1",
-                                        #     "resource": "authorizationpolicies",
-                                        #     "updateStrategy": {"method": "InPlace"},
-                                        # },
-                                    ],
-                                    "generateSelector": True,
-                                    "hooks": {
-                                        "sync": {
-                                            "webhook": {
-                                                "url": f"http://"
-                                                f"{self.model.app.name}.{self.model.name}/sync"
-                                            }
-                                        }
-                                    },
-                                    "parentResource": {
-                                        "apiVersion": "v1",
-                                        "resource": "namespaces",
-                                    },
-                                    "resyncPeriodSeconds": 3600,
-                                },
-                            }
-                        ]
-                    },
-                    "secrets": [
-                        {
-                            "name": f"{self.model.app.name}-minio-credentials",
-                            "type": "Opaque",
-                            "data": {
-                                k: b64encode(v.encode("utf-8")).decode("utf-8")
-                                for k, v in {
-                                    "MINIO_ACCESS_KEY": os["access-key"],
-                                    "MINIO_SECRET_KEY": os["secret-key"],
-                                }.items()
-                            },
-                        }
+                inputs_getter=lambda: KfpProfileControllerInputs(
+                    MINIO_SECRET=json.dumps(
+                        {"secret": {"name": f"{self.model.app.name}-minio-credentials"}}
+                    ),
+                    MINIO_HOST=self.object_storage_relation.component.get_data()["service"],
+                    MINIO_PORT=self.object_storage_relation.component.get_data()["port"],
+                    MINIO_NAMESPACE=self.object_storage_relation.component.get_data()["namespace"],
+                    MINIO_ACCESS_KEY=self.object_storage_relation.component.get_data()[
+                        "access-key"
```

Comment:
Reading this here, I realise we're doing a lot of independent `get_data()` calls.  Do you think it is worth making this into a real function and calling `.get_data()` just once?  I have the same problem in kfp-ui too

Reply:
You can't its a lambda which will be called inside the component .... soo I cannot precached the values .... at least I think :D